File: c/immutable_text.cc
#include<typeinfo>
#include<string.h>
#include <string>
#include <iostream>

using namespace std;

// <summary>Holds the default size for primitive blocks of characters.</summary>
const int BLOCK_SIZE = 1 << 6;

/// <summary>Holds the mask used to ensure a block boundary cesures.</summary>
const int BLOCK_MASK = ~(BLOCK_SIZE - 1);

class Node {
public:
		virtual int Length() = 0;
		virtual Node* SubNode(int start, int end) = 0;
		virtual wchar_t GetCharAt(int index) = 0;
		virtual void CopyTo (int sourceIndex, wchar_t* destination, int destinationIndex, int count) = 0;
};

class WideLeafNode : public Node
{
private:
	int size;
	const wchar_t* data;
public:
	WideLeafNode(const wchar_t* data, int size)
	{
		this->data = data;
		this->size = size;
	}

	int Length() {
		return size;
	}

	virtual Node* SubNode(int start, int end)  {
		if (start == 0 && end == Length()) {
			return this;
		}
		wchar_t* subArray = new wchar_t[end - start];
		CopyTo(start, subArray, 0, end - start);
		return new WideLeafNode(subArray, end - start);
	}

	wchar_t GetCharAt(int index)
	{
		return *(data + index);
	}

	void CopyTo (int sourceIndex, wchar_t* destination, int destinationIndex, int count) {
		const wchar_t* src = data;
		wchar_t* ptr = destination + destinationIndex;
		wchar_t* end = ptr + count;
		while(ptr != end) {
			*ptr = *src;
			ptr++;src++;
		}
	}

	~WideLeafNode()
	{
		delete [] data;
	}
};

static Node* NodeOf (Node* node, int offset, int length);
static Node* ConcatNodes (Node* node1, Node* node2);

class CompositeNode : public Node
{
public:
	int count;
	Node* head;
	Node* tail;

	CompositeNode(int count, Node* head, Node* tail)
	{
		this->count = count;
		this->head = head;
		this->tail = tail;
	}

	int Length() {
		return count;
	}

	virtual Node* SubNode(int start, int end)  {
		int cesure = head->Length();
		if (end <= cesure) {
			return head->SubNode(start, end);
		}

		if (start >= cesure) {
		 return tail->SubNode(start - cesure, end - cesure);
	 }

		if (start == 0 && end == count)
			return this;
		// Overlaps head and tail.
		return ConcatNodes(head->SubNode(start, cesure), tail->SubNode(0, end - cesure));
	}

	wchar_t GetCharAt(int index)
	{
		int headLength = head->Length();
		if (index < headLength) {
			return head->GetCharAt(index);
		}
		return tail->GetCharAt(index - headLength);
	}

	void CopyTo (int sourceIndex, wchar_t* destination, int destinationIndex, int count) {
		int cesure = head->Length();
		if (sourceIndex + count <= cesure) {
			head->CopyTo(sourceIndex, destination, destinationIndex, count);
			return;
		}
		if (sourceIndex >= cesure) {
			tail->CopyTo(sourceIndex - cesure, destination, destinationIndex, count);
			return;
		}
		// Overlaps head and tail.
		int headChunkSize = cesure - sourceIndex;
		head->CopyTo(sourceIndex, destination, destinationIndex, headChunkSize);
		tail->CopyTo(0, destination, destinationIndex + headChunkSize, count - headChunkSize);
	}

	~CompositeNode()
	{
		delete head;
		delete tail;
	}

	CompositeNode* RotateRight()
	{
		if (typeid(*head) != typeid (CompositeNode))
			return this;
		CompositeNode* P = (CompositeNode*)head;
		Node* A = P->head;
	 	Node* B = P->tail;
	 	Node* C = this->tail;
		int tailLength = B->Length() + C->Length();
		return new CompositeNode(A->Length() + tailLength, A, new CompositeNode(tailLength, B, C ));
	}

	CompositeNode* RotateLeft()
	{
	 // See: http://en.wikipedia.org/wiki/Tree_rotation
	 if (typeid(*tail) != typeid (CompositeNode))
		 return this;

		 CompositeNode* Q = (CompositeNode*)tail;
	 Node* B = Q->head;
	 Node* C = Q->tail;
	 Node* A = this->head;
	 int headLength = A->Length() + B->Length();
	 return new CompositeNode (headLength +  C->Length(), new CompositeNode (headLength, A, B), C);
	}
};

static Node* ConcatNodes (Node* node1, Node* node2)
{
 // All Text instances are maintained balanced:
 //   (head < tail * 2) & (tail < head * 2)
 int length = node1->Length() + node2->Length();
 if (length <= BLOCK_SIZE) { // Merges to primitive.
	 wchar_t* mergedArray = new wchar_t[length];
	 node1->CopyTo (0, mergedArray, 0, node1->Length());
	 node2->CopyTo (0, mergedArray, node1->Length(), node2->Length());
	 return new WideLeafNode(mergedArray, length);
 }
 // Returns a composite.
 Node* head = node1;
 Node* tail = node2;
 CompositeNode* compositeTail = (CompositeNode*)tail;
 if (typeid(*tail) == typeid (CompositeNode) && (head->Length() << 1) < tail->Length()) {
	 // head too small, returns (head + tail/2) + (tail/2)
	 if (compositeTail->head->Length() > compositeTail->tail->Length()) {
		 // Rotates to concatenate with smaller part.
		 compositeTail = compositeTail->RotateRight ();
	 }
	 head = ConcatNodes (head, compositeTail->head);
	 tail = compositeTail->tail;
 } else {
	 if (typeid(*head) == typeid (CompositeNode)) {
		 CompositeNode* compositeHead = (CompositeNode*)head;
		 if ((tail->Length() << 1) < head->Length()) {
			 // tail too small, returns (head/2) + (head/2 concat tail)
			 if (compositeHead->tail->Length() > compositeHead->head->Length()) {
				 // Rotates to concatenate with smaller part.
				 compositeHead = compositeHead->RotateLeft ();
			 }
			 tail = ConcatNodes (compositeHead->tail, tail);
			 head = compositeHead->head;
		 }
	 }
 }
 return new CompositeNode (head->Length() + tail->Length(), head, tail);
}

static Node* NodeOf (Node* node, int offset, int length)
{
 if (length <= BLOCK_SIZE) {
	 return node->SubNode (offset, offset + length);
 }
 // Splits on a block boundary.
 int half = ((length + BLOCK_SIZE) >> 1) & BLOCK_MASK;
 Node* head = NodeOf (node, offset, half);
 Node* tail = NodeOf (node, offset + half, length - half);
 return new CompositeNode (head->Length() + tail->Length(), head, tail);
}

struct InnerLeaf
{
	Node* leafNode;
	int offset;
};

class ImmutableText
{
private:
	Node* root;
public:
	ImmutableText(Node* root)
	{
		this->root = root;
	}

	ImmutableText(wstring str)
	{
		wchar_t* runes = new wchar_t[str.length()];
		memcpy(runes, str.c_str(), str.length() * sizeof(wchar_t));
		this->root = new WideLeafNode(runes, str.length());
	}

	int Length()
	{
		return root->Length();
	}

	wchar_t GetCharAt(int index)
	{
		InnerLeaf leaf = FindLeaf(index, 0);
		return leaf.leafNode->GetCharAt(index - leaf.offset);
	}

	ImmutableText* EnsureChunked()
	{
		int len = Length();
		if (typeid(*root) != typeid (CompositeNode) || len > BLOCK_SIZE)
			return new ImmutableText(NodeOf (root, 0, len));
		return this;
	}

	InnerLeaf FindLeaf(int index, int offset)
	{
		InnerLeaf result;
		result.leafNode = 0;
		result.offset = -1;
		Node* node = this->root;
		while (1) {
			if (index >= node->Length()) {
				return result;
			}
			if (typeid(*node) == typeid (CompositeNode)) {
				CompositeNode* composite = (CompositeNode*)node;
				if (index < composite->head->Length()) {
					node = composite->head;
				} else {
					offset += composite->head->Length();
					index -= composite->head->Length();
					node = composite->tail;
				}
				continue;
			}
			result.leafNode = node;
			result.offset = offset;
			return result;
		}
	}

	/// <summary>
	/// Concatenates the specified text to the end of this text.
	/// This method is very fast (faster even than
	/// <code>StringBuffer.append(String)</code>) and still returns
	/// a text instance with an internal binary tree of minimal depth!
	/// </summary>
	/// <param name="that">that the text that is concatenated.</param>
	/// <returns><code>this + that</code></returns>
	ImmutableText* Concat(ImmutableText* that)
	{
		if (that->Length() == 0) {
			return this;
		}
		if (this->Length() == 0) {
			return that;
		}
		return new ImmutableText (ConcatNodes (this->EnsureChunked()->root, that->EnsureChunked()->root));
	}

	/// <summary>
	/// Returns a portion of this text.
	// </summary>
	/// <returns>the sub-text starting at the specified start position and ending just before the specified end position.</returns>
	ImmutableText* GetText(int start, int count)
	{
		int end = start + count;
		if (start == 0 && end == Length()) {
			return this;
		}
		if (start == end) {
			return new ImmutableText (new WideLeafNode (new wchar_t[0], 0));
		}
		return new ImmutableText (this->root->SubNode (start, end));
	}

	ImmutableText* InsertText(int index, ImmutableText* text)
	 {
		 // TODO: Fix memory leaking - since it's possible that these methods return 'this' delete always will not work here.
		return this->GetText (0, index)->Concat (text)->Concat (this->SubText (index));
	}
	/*
	func (this ImmutableText) InsertString(index int, text string) ImmutableText {
		return this.InsertText (index, CreateImmutableText (text))
	}

	func CreateImmutableText (text string) ImmutableText {
		return ImmutableText { WideLeafNode { []rune(text) } }
	}

	*/

	/// <summary>
	/// Returns the text without the characters between the specified indexes.
	/// </summary>
	/// <returns><code>subtext(0, start).concat(subtext(end))</code></returns>
	ImmutableText* RemoveText(int start, int count)
	{
		if (count == 0)
			return this;
		int end = start + count;
	//	if (end > Length)
	//		throw new IndexOutOfRangeException ();
		// TODO: Fix memory leaking - since it's possible that these methods return 'this' delete always will not work here.
		return this->EnsureChunked ()->GetText (0, start)->Concat (this->SubText (end));
	}

	ImmutableText* SubText(int start)
	{
		return this->GetText (start, this->Length() - start);
	}
	wstring ToString()
	{
		wchar_t* runes = new wchar_t[this->Length()];
		this->root->CopyTo(0, runes, 0, this->Length());
		return wstring(runes);
	}
};

int main() {
	ImmutableText* insert = new ImmutableText(L"1");
	for (int j = 0; j < 100; j++) {
		ImmutableText* myText = new ImmutableText(L"hello");
		for (int i = 0; i < 1000; i++) {
			myText = myText->InsertText(i, insert);
		}
		for (int i = 0; i < 1000; i++) {
			myText = myText->RemoveText(0, 1);
		}
	}

	return 0;
}

File: rust/ImmutableText.rs
use std::cmp;

// <summary>Holds the default size for primitive blocks of characters.</summary>
const BLOCK_SIZE : i32 = 1 << 6;

/// <summary>Holds the mask used to ensure a block boundary cesures.</summary>
const BLOCK_MASK : i32 = !(BLOCK_SIZE - 1);

trait Node {
	fn length(&self) -> usize;
	fn sub_node(&self, start: usize, end: usize) -> Node;
	fn get_char_at(&self, offset : usize) -> char;
	fn copy_to(&self, source_index : usize, destination : &mut [char], destination_index : usize, count : usize);
}

struct WideLeafNode {
	data:Vec<char>
}

impl Node for WideLeafNode {
	fn length(&self) -> usize {
		self.data.len()
	}
	
	fn sub_node<'a>(&self, start: usize, end: usize) -> Node {
		if start == 0 && end == self.length() {
			return self;
		}
		let mut vec = Vec::with_capacity(end - start);
		let mut i = 0;
		for j in start..end {
			vec[i] = self.data[j];
			i = i + 1;
		}
		let mut result = WideLeafNode {
			data : vec
		};

		result
	}
	
	fn get_char_at(&self, offset : usize) -> char {
		self.data[offset]
	}

	fn copy_to(&self, source_index : usize, destination : &mut [char], destination_index : usize, count : usize) {
		for i in 0..count {
			destination[destination_index + i] = self.data[source_index +i];
		}
	}
}



/*
func (this WideLeafNode) CopyTo(sourceIndex int, destination []rune, destinationIndex int, count int) {
	copy(destination[destinationIndex:], this.data[sourceIndex:count])
}

type CompositeNode struct {
	count int
	head Node
	tail Node
}

func (this CompositeNode) Length() int {
	return this.count
}

func (this CompositeNode) SubNode(start int, end int) Node {
	var cesure = this.head.Length()
	if end <= cesure {
		return this.head.SubNode (start, end)
	}
	if start >= cesure {
		return this.tail.SubNode (start - cesure, end - cesure)
	}
	if start == 0 && end == this.count {
		return this
	}
	// Overlaps head and tail.
	return ConcatNodes (this.head.SubNode (start, cesure), this.tail.SubNode (0, end - cesure))
}

func (this CompositeNode) GetCharAt(index int) rune {
	var headLength = this.head.Length()
	if index < headLength {
		return this.head.GetCharAt(index)
	}
	return this.tail.GetCharAt(index - headLength)
}

func (this CompositeNode) CopyTo(sourceIndex int, destination []rune, destinationIndex int, count int) {
	var cesure = this.head.Length ()
	if sourceIndex + count <= cesure {
		this.head.CopyTo (sourceIndex, destination, destinationIndex, count)
		return
	}
	if (sourceIndex >= cesure) {
		this.tail.CopyTo (sourceIndex - cesure, destination, destinationIndex, count)
		return
	}
	// Overlaps head and tail.
	var headChunkSize = cesure - sourceIndex;
	this.head.CopyTo (sourceIndex, destination, destinationIndex, headChunkSize)
	this.tail.CopyTo (0, destination, destinationIndex + headChunkSize, count - headChunkSize)
}

func (this CompositeNode) RotateRight () CompositeNode {
	// See: http://en.wikipedia.org/wiki/Tree_rotation
	var P, isType = this.head.(CompositeNode);
	if !isType {
		return this // Head not a composite, cannot rotate.
	}
	var A = P.head
	var B = P.tail
	var C = this.tail
	var tailLength = B.Length() + C.Length ()
	return CompositeNode { A.Length () + tailLength, A, CompositeNode { tailLength, B, C } }
}

func (this CompositeNode) RotateLeft () CompositeNode {
	// See: http://en.wikipedia.org/wiki/Tree_rotation
	var Q, isType = this.tail.(CompositeNode)
	if !isType {
		return this // Tail not a composite, cannot rotate.
	}
	var B = Q.head
	var C = Q.tail
	var A = this.head
	var headLength = A.Length() + B.Length()
	return CompositeNode { headLength +  C.Length(), CompositeNode { headLength, A, B }, C }
}

func NodeOf (node Node, offset int, length int) Node {
	if (length <= BLOCK_SIZE) {
		return node.SubNode (offset, offset + length);
	}
	// Splits on a block boundary.
	var half = ((length + BLOCK_SIZE) >> 1) & BLOCK_MASK
	var head = NodeOf (node, offset, half)
	var tail = NodeOf (node, offset + half, length - half)
	return CompositeNode { head.Length() + tail.Length(), head, tail }
}

func ConcatNodes (node1 Node, node2 Node) Node  {
	// All Text instances are maintained balanced:
	//   (head < tail * 2) & (tail < head * 2)
	var length = node1.Length() + node2.Length()
	if length <= BLOCK_SIZE { // Merges to primitive.
		var mergedArray = make ([]rune, length)
		node1.CopyTo (0, mergedArray, 0, node1.Length())
		node2.CopyTo (0, mergedArray, node1.Length(), node2.Length())
		return CreateLeafNode (mergedArray)
	}
	// Returns a composite.
	var head = node1
	var tail = node2
	var compositeTail, isType = tail.(CompositeNode)
	if (head.Length() << 1) < tail.Length() && isType {
		// head too small, returns (head + tail/2) + (tail/2)
		if compositeTail.head.Length() > compositeTail.tail.Length() {
			// Rotates to concatenate with smaller part.
			compositeTail = compositeTail.RotateRight ()
		}
		head = ConcatNodes (head, compositeTail.head)
		tail = compositeTail.tail
	} else {
		var compositeHead, isType2 = head.(CompositeNode)
		if (tail.Length() << 1) < head.Length() && isType2 {
			// tail too small, returns (head/2) + (head/2 concat tail)
			if (compositeHead.tail.Length() > compositeHead.head.Length()) {
				// Rotates to concatenate with smaller part.
				compositeHead = compositeHead.RotateLeft ()
			}
			tail = ConcatNodes (compositeHead.tail, tail)
			head = compositeHead.head
		}
	}
	return CompositeNode { head.Length() + tail.Length(), head, tail }
}

func CreateLeafNode (str []rune) Node {
//			byte [] bytes = ToBytesIfPossible (str);
//			if (bytes != null)
//				return new Leaf8BitNode (bytes);
	return WideLeafNode { str}
}

type ImmutableText struct {
	root Node
}

func (this ImmutableText) Length() int {
	return this.root.Length()
}

func (this ImmutableText) GetCharAt(index int) rune {
	var leaf = this.FindLeaf(index, 0)
	return (*leaf.leafNode).GetCharAt(index - leaf.offset)
}

type InnerLeaf struct {
	leafNode *Node
	offset int
}

func (this ImmutableText) EnsureChunked() ImmutableText {
	var len = this.Length()
	var composite, isComposite = this.root.(CompositeNode)
	if len > BLOCK_SIZE && !isComposite {
		return ImmutableText { NodeOf (composite, 0, len) }
	}
	return this
}

func (this ImmutableText) FindLeaf(index int, offset int) InnerLeaf {
	var node = this.root
	for {
		if index >= node.Length() {
			return InnerLeaf { nil, -1 }
		}

		var composite, isComposite = node.(CompositeNode)
		if isComposite {
			if (index < composite.head.Length()) {
				node = composite.head
			} else {
				offset += composite.head.Length()
				index -= composite.head.Length()
				node = composite.tail
			}
			continue
		}

		return InnerLeaf { &node, offset }
	}
}

/// <summary>
/// Concatenates the specified text to the end of this text.
/// This method is very fast (faster even than
/// <code>StringBuffer.append(String)</code>) and still returns
/// a text instance with an internal binary tree of minimal depth!
/// </summary>
/// <param name="that">that the text that is concatenated.</param>
/// <returns><code>this + that</code></returns>
func (this ImmutableText) Concat(that ImmutableText) ImmutableText {
	if that.Length() == 0 {
		return this
	}
	if this.Length() == 0 {
		return that
	}
	return ImmutableText { ConcatNodes (this.EnsureChunked().root, that.EnsureChunked().root) }
}

/// <summary>
/// Returns a portion of this text.
// </summary>
/// <returns>the sub-text starting at the specified start position and ending just before the specified end position.</returns>
func (this ImmutableText) GetText(start int, count int) ImmutableText {
	var end = start + count
//	if ((start < 0) || (start > end) || (end > Length)) {
//		throw new IndexOutOfRangeException (" start :" + start + " end :" + end + " needs to be between 0 <= " + Length)
//	}
	if start == 0 && end == this.Length() {
		return this
	}
	if start == end {
		return ImmutableText { WideLeafNode {make([]rune, 0)}}
	}
	return ImmutableText { this.root.SubNode (start, end) }
}

func (this ImmutableText) InsertText(index int, text ImmutableText) ImmutableText {
	return this.GetText (0, index).Concat (text).Concat (this.SubText (index))
}

func (this ImmutableText) InsertString(index int, text string) ImmutableText {
	return this.InsertText (index, CreateImmutableText (text))
}

func CreateImmutableText (text string) ImmutableText {
	return ImmutableText { WideLeafNode { []rune(text) } }
}

/// <summary>
/// Returns the text without the characters between the specified indexes.
/// </summary>
/// <returns><code>subtext(0, start).concat(subtext(end))</code></returns>
func (this ImmutableText) RemoveText(start int, count int) ImmutableText {
	if count == 0 {
		return this
	}
	var end = start + count
//	if (end > Length)
//		throw new IndexOutOfRangeException ();
	return this.EnsureChunked ().GetText (0, start).Concat (this.SubText (end))
}

func (this ImmutableText) SubText(start int) ImmutableText {
	return this.GetText (start, this.Length() - start)
}

func (this ImmutableText) ToString() string {
	var runes = make([]rune, this.Length())
	this.root.CopyTo(0, runes, 0, this.Length())
	return string(runes)
}
*/

fn main() {

  println!("Hello world");

  /*
	for j := 0; j < 100; j++ {
		var myText = CreateImmutableText("hello")
		for i := 0; i < 100000; i++ {
			myText = myText.InsertString(i, "1")
		}
		for i := 0; i < 100000; i++ {
			myText = myText.RemoveText(0, 1)
		}
	}
  */
}
